src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java

```diff
@@ -187,53 +193,95 @@ private void doStart() throws Exception {
 
     }
 
-    /** Similar to {@code MaskPasswordsOutputStream}. */
-    private static final class Filter extends ConsoleLogFilter implements Serializable {
-
+    private static class SecretPattern implements Serializable {
         private static final long serialVersionUID = 1;
 
-        private final Secret pattern;
-        private String charsetName;
-        
-        Filter(Collection<String> secrets, String charsetName) {
-            pattern = Secret.fromString(SecretPatterns.getAggregateSecretPattern(secrets).pattern());
-            this.charsetName = charsetName;
+        private transient final Pattern pattern;
+
+        private SecretPattern(Pattern pattern) {
+            this.pattern = pattern;
         }
-        
-        // To avoid de-serialization issues with newly added field (charsetName)
+
+        Matcher matcher(CharSequence input) {
+            return pattern.matcher(input);
+        }
+
+        private Object writeReplace() {
+            return new SerializationProxy(Secret.fromString(pattern.pattern()));
+        }
+
         private Object readResolve() throws ObjectStreamException {
-            if (this.charsetName == null) {
-                this.charsetName = Charsets.UTF_8.name();
+            throw new InvalidObjectException("Must use proxy");
+        }
+
+        private static class SerializationProxy implements Serializable {
```

Comment:
Seems like overkill. Just have `SecretPattern` compile its pattern on demand.

Reply:
You just have

```java
private synchronized Pattern pattern() {
  if (pattern == null) {
    pattern = Pattern.compile(secret);
  }
  return pattern;
}
```

so it is compiled at most once per actual instance of `SecretPattern`.

Reply:
Then I think this class is only here due to best practices with serialization (i.e., use a serialization proxy unless your class is a pure data POJO of some sort with no enforced invariants).

Reply:
Just suggesting a simplification that avoids the need for `writeReplace` / `readResolve` tricks.